src/liballoc/collections/btree/set.rs

```diff
@@ -944,6 +945,36 @@ impl<T: Ord> BTreeSet<T> {
     {
         BTreeSet { map: self.map.split_off(key) }
     }
+
+    /// Creates an iterator which uses a closure to determine if a value should be removed.
+    ///
+    /// If the closure returns true, then the value is removed and yielded.
+    /// If the closure returns false, the value will remain in the list and will not be yielded
+    /// by the iterator.
+    ///
+    /// # Examples
+    ///
+    /// Splitting a set into even and odd values, reusing the original set:
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(btree_drain_filter)]
+    /// use std::collections::BTreeSet;
+    ///
+    /// let mut set: BTreeSet<i32> = (0..8).collect();
+    /// let evens: BTreeSet<_> = set.drain_filter(|v| v % 2 == 0).collect();
+    /// let odds = set;
+    /// assert_eq!(evens.into_iter().collect::<Vec<_>>(), vec![0, 2, 4, 6]);
+    /// assert_eq!(odds.into_iter().collect::<Vec<_>>(), vec![1, 3, 5, 7]);
+    /// ```
+    #[unstable(feature = "btree_drain_filter", issue = "42849")]
+    pub fn drain_filter<'a, F>(&'a mut self, mut pred: F) -> DrainFilter<'a, T>
+    where
+        F: 'a + FnMut(&T) -> bool,
+    {
+        DrainFilter(self.map.drain_filter(Box::new(move |k, _| pred(k))))
```

Comment:
I do see why you're erasing the type into a `Box`, with that unnameable internal closure type, but this isn't desirable. At the very least, the type signatures should still look like we're using `F` directly.

Reply:
I had [a question about that](https://users.rust-lang.org/t/can-we-generically-wrap-a-struct-with-closure-field/35476), and it seemed there was no alternative. I don't know which type signatures you mean there and definitely not how to do that.

Reply:
I think I know now what type signatures you meant.